pkg/il/program.go

```diff
@@ -0,0 +1,160 @@
+// There is no constant representation of Record in il form.
+//  - Void represents no-value. This is typically used in the return type of a function, to indicate
+// that the function does not return a result.
+//
+// In addition to these types, the VM inherently supports errors, based on Go errors. It is possible
+// to raise these errors from within the IL. However, it is not possible to trap or return these as
+// values. Once an error is raised, whether as a program instruction, or due to an invalid evaluation
+// step, it bubbles all the way out of the evaluation context.
+//
+// The execution model is based on an operand stack and registers. Though most operations work on an
+// operand stack to perform calculations, it is possible to transfer data from registers to stack
+// and vice-versa.
+//
+// The IL supports a stack-based calling-convention. All the parameters to a function needs to be
+// pushed onto the stack before CALL instruction is invoked. Similarly, the return values needs to be
+// pushed onto the stack when RET is called.
+//
+// Typically, an IL based program is created by a compiler by initializing a Program type and
+// by adding Functions to it, whose bodies are built using the Builder type. Once built, programs
+// can be serialized/deserialized into a textual form for ease of use, using Write* and Read* method
+// in this package.
+//
+package il
+
+import (
+	"fmt"
+)
+
+const (
+	// programCodeSize is the default size of a program.
+	programCodeSize int = 256
+)
+
+// Program is a self-contained IL based set of functions that can be executed by an interpreter.
+type Program struct {
+	// Strings is the collection of strings that is used by the program.
+	strings *StringsTable
+
+	// Functions is the collection of functions that are contained in the program.
+	Functions *FunctionsTable
+
+	// Code is the actual byte-code based body of all the functions in the program.
+	code []uint32
+
+	// codeEnd points to the end of the byte-codes within code.
+	codeEnd uint32
+}
+
+// NewProgram creates and returns a new, empty program.
+func NewProgram() Program {
+	strings := newStringsTable()
+	p := Program{
+		strings:   strings,
+		Functions: newFunctionsTable(strings),
+		code:      make([]uint32, programCodeSize),
+		codeEnd:   1,
+	}
+
+	p.code[0] = uint32(Halt)
+
+	return p
+}
+
+// AddExternDef adds a new external/native function to the program. The actual bodies of functions
+// needs to be supplied to the interpreter separately.
+func (p *Program) AddExternDef(name string, parameters []Type, returnType Type) {
+	f := Function{
+		ID:         p.strings.GetID(name),
+		Length:     0,
+		Address:    0,
+		Parameters: parameters,
+		ReturnType: returnType,
+	}
+	p.Functions.add(&f)
+}
+
+// AddFunction adds a new IL based function to the program.
+func (p *Program) AddFunction(name string, parameters []Type, returnType Type, body []uint32) error {
```

Comment:
That would require more casts. Opcode type is only for the opcode itself, not for the arguments. Body contains both.